src/connectors/fname/mod.rs

```diff
@@ -0,0 +1,105 @@
+use tokio::time::{sleep, Duration};
+use serde::Deserialize;
+use thiserror::Error;
+use tracing::{debug, info, warn, span, Level, Span};
+
+
+#[derive(Deserialize, Debug)]
+struct TransfersData {
+    transfers: Vec<Transfer>,
+}
+
+#[derive(Deserialize, Debug)]
+struct Transfer {
+    id: u64,
+    timestamp: u64,
+    username: String,
+    owner: String,
+    from: u64,
+    to: u64,
+    user_signature: String,
+    server_signature: String,
+}
+
+
+#[derive(Error, Debug)]
+pub enum FnamesError {
+    #[error("non-sequential IDs found")]
+    NonSequentialIds,
+}
+
+
+#[derive(Error, Debug)]
+enum FetchError {
+    #[error("non-sequential IDs found")]
+    NonSequentialIds,
+
+    #[error("no new IDs found")]
+    NoNewIDs,
+
+    #[error(transparent)]
+    Reqwest(#[from] reqwest::Error),
+}
+
+
+pub struct Fetcher {
+    position: u64,
+    transfers: Vec<Transfer>,
+}
+
+
+impl Fetcher {
+    pub fn new(start_from: u64) -> Self {
+        Fetcher {
+            position: start_from,
+            transfers: vec![],
+        }
+    }
+
+    async fn fetch(&mut self) -> Result<u64, FetchError> {
+        loop {
+            let url = format!("https://fnames.farcaster.xyz/transfers?from_id={}", self.position);
+            debug!(%url, "fetching transfers");
+
+            let response = reqwest::get(&url)
+                .await?
+                .json::<TransfersData>()
+                .await?;
+
+            let count = response.transfers.len();
+
+            if count == 0 {
+                return Ok(self.position);
+            }
+
+            info!(count, position=self.position, "found new transfers");
+
+            for t in response.transfers {
+                if t.id <= self.position {
+                    return Err(FetchError::NonSequentialIds);
+                }
+                self.position = t.id;
+                self.transfers.push(t); // Just store these for now, we'll use them later
+            }
+        }
+    }
+
+    pub async fn run(&mut self) -> FnamesError {
+        loop {
+            match self.fetch().await {
+                Ok(new_count) => self.position = new_count,
+                Err(e) => match e {
+                    FetchError::NonSequentialIds => {
+                        return FnamesError::NonSequentialIds;
```

Comment:
Can we use the same type everywhere?

Also, even in this scenario, we don't want the fnames fetcher to stop. Because that would mean the chain cannot ingest new dependent messages or it might require a restart of all the nodes. As a general principle, the nodes should prioritize uptime/liveness. 

This would also affect consensus (and cause chain halts) because if only certain validators see this error and stop fetching fnames, then they cannot validate fnames included by other validators that did not encounter this.

In practice, this error is unlikely to ever happen so it's not a big deal, but it's worth thinking through these things because these "consensus critical" issues are going to be the biggest source of pain.

Reply:
Yes, definitely worth thinking through!

With above context, agree that retrying is probably better. If we got non-sequential IDs, it probably means that something went really off the rails, and if it persists we'll probably break consensus and eventually halt production. But perhaps something resolves on the fname server in a retry and we're back in business.

> Can we use the same type everywhere?

Did you mean the fact that there was both NonSequentialIds in both FetchError and FnamesError? I'm not sure, this is my lack of rust experience shining through. I'd like to know what the best solution here is. This was the least ugly configuration I could come up with. But also since we'll be retrying instead of returning FnamesError, this issue will no longer exist.